core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala

```diff
@@ -0,0 +1,258 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler
+
+import java.util.concurrent.{ExecutorService, TimeUnit, ConcurrentHashMap}
+
+import scala.collection.{Map => ScalaImmutableMap}
+import scala.collection.convert.decorateAsScala._
+
+import akka.actor.{ActorRef, Actor}
+
+import org.apache.spark.{SparkConf, Logging}
+import org.apache.spark.util.{Utils, AkkaUtils, ActorLogReceive}
+
+private[spark] sealed trait OutputCommitCoordinationMessage
+
+private[spark] case class StageStarted(stage: Int, partitionIds: Seq[Int])
+  extends OutputCommitCoordinationMessage
+private[spark] case class StageEnded(stage: Int) extends OutputCommitCoordinationMessage
+private[spark] case object StopCoordinator extends OutputCommitCoordinationMessage
+
+private[spark] case class AskPermissionToCommitOutput(
+    stage: Int,
+    task: Long,
+    partId: Int,
+    taskAttempt: Long)
+  extends OutputCommitCoordinationMessage with Serializable
+
+private[spark] case class TaskCompleted(
+    stage: Int,
+    task: Long,
+    partId: Int,
+    attempt: Long,
+    successful: Boolean)
+  extends OutputCommitCoordinationMessage
+
+/**
+ * Authority that decides whether tasks can commit output to HDFS.
+ *
+ * This lives on the driver, but the actor allows the tasks that commit
+ * to Hadoop to invoke it.
+ */
+private[spark] class OutputCommitCoordinator(conf: SparkConf) extends Logging {
+
+  private type StageId = Int
+  private type PartitionId = Int
+  private type TaskId = Long
+  private type TaskAttemptId = Long
+
+  // Wrapper for an int option that allows it to be locked via a synchronized block
+  // while still setting option itself to Some(...) or None.
+  private class LockableAttemptId(var value: Option[TaskAttemptId])
+
+  private type CommittersByStageHashMap =
```

Comment:
I would be very surprised if the thread pool resulted in a significant performance improvement.  The extra overheads of submitting tasks to the pool, working with immutable maps, etc. might actually harm performance unless we recover other time due to the parallelism.

I'm not super-familiar with Akka internals, but the fact that an actor processes messages serially doesn't necessarily mean that the lower-level message transport code needs to be single-threaded.

Until we know that this gives a performance improvement or we observe a performance regression due to this patch, I'd like to roll back these thread-safety and parallelism changes and stick with a simpler single-threaded implementation; I think it will be easier to understand and less bug-prone.